File: 1350-remove-sub-folders-from-the-filesystem/remove-sub-folders-from-the-filesystem.py
class Trie:
    def __init__(self):
        self.children = {}
        self.endOfWord = False
    
    def add(self, word):
        curr = self
        for i in word.split('/'):
            if i not in curr.children:
                curr.children[i] = Trie()
            curr = curr.children[i]
        curr.endOfWord = True

    def prefixSearch(self, word):
        curr = self
        for i in word.split('/')[:-1]:
            curr = curr.children[i]
            if curr.endOfWord:
                return True
        return False

class Solution:
    def removeSubfolders(self, folder: List[str]) -> List[str]:
        curr = Trie()

        result = []
        for i in folder:
            curr.add(i)
        
        for i in folder:
            if  curr.prefixSearch(i):
                continue
            result.append(i)

        return result


        # cache = Trie()
        # result = []

        # for f in folder:
        #     temp = cache
        #     for d in f[1:].split('/'):
        #         if d not in temp.childrens:
        #             temp.childrens[d] = Trie() 
        #         temp = temp.childrens[d] 
        #     temp.end = True
        
        

        # stack = [(cache, "")]
        # while stack:
        #     temp1, temp2 = stack.pop(0)
        #     if temp1.end:
        #         result.append(temp2)
        #         continue
        #     for i, j in temp1.childrens.items():
        #         stack.append((j, temp2+'/'+i))


            

        # return result

        # cache = set(folder)
        # result = []
        # for i in folder:
        #     result.append(i)
        #     for j in range(len(i)):
        #         if i[j] == "/" and i[:j] in cache:
        #             result.pop()
        #             break
        # return result

File: 1350-remove-sub-folders-from-the-filesystem/remove-sub-folders-from-the-filesystem.java
class Trie{
    HashMap<String, Trie> children = new HashMap<String, Trie>();
    boolean endOfWord = false;

    void add(String s){
        Trie curr = this;
        for (String ss: s.split("/")){
            if (!curr.children.containsKey(ss)){
                curr.children.put(ss, new Trie());
            }
            curr = curr.children.get(ss);
        }
        curr.endOfWord = true;
    }

    boolean prefix(String s){
        Trie curr = this;
        String[] splits = s.split("/");
        
        for (int i = 0; i < splits.length - 1; i++){
            curr = curr.children.get(splits[i]);
            if (curr.endOfWord) return true;
        }
        return false;
    }
}
class Solution {
    public List<String> removeSubfolders(String[] folder) {
        Trie cache = new Trie();
        List<String> result = new ArrayList<String>();
        for (String t: folder) cache.add(t);

        for (String t: folder){
            if (! cache.prefix(t)) result.add(t);
        }
        return result;
    }
    // public List<String> removeSubfolders(String[] folder) {
    //     List<String> result = new ArrayList<String>();
    //     HashSet<String> cache = new HashSet<String>();
    //     for (String i: folder) cache.add(i);
    //     for (String i: folder){
    //         result.add(i);
    //         for (int j = 0; j < i.length(); j++){
    //             if (i.charAt(j) == '/' && cache.contains(i.substring(0, j))){
    //                 result.remove(result.size() -1 );
    //                 break;
    //             }
    //         }
    //     }
    //     return result;
    // }
}